### PR TITLE
FSI - Work in progress to get a working version

### PR DESCRIPTION
**Pull request by gantech**
Ready for merge if clears formatting and other checks

### CODE REVIEW

File: src/Realm.C [hunk 1]
```diff
@@ -63,6 +63,7 @@
 
 // actuator line/fsi
 #include <aero/AeroContainer.h>
+#include <aero/fsi/OpenfastFSI.h>
```
Comment (psakievich) [inline src/Realm.C:66]:
```suggestion
```
This header should not be needed.  All dependencies on FSI code should be isolated to the AeroContainer

**Comment by psakievich** [comment]
I've tested this on device.  It adds several warnings from the master element functions we are using on host that are `__host__ __device` functions, but it is okay since we are only using them on host.  

Also the openfast based unit tests still segfault using the fsi branch, but we already knew that.  Something to track down still. 

We should also delete the extra print statements, but since we are going to refactor this soon and this code is all isolated unless doing FSI work I'm inclined to leave them for now.

Those seem to be the only outstanding issues and I don't think they should keep us from merging.

**Review by psakievich — APPROVED** [review]

**Comment by gantech** [comment]
Thanks Phil. @ashesh2512 Could you please go over this and see if we missed anything?

File: src/aero/fsi/FSIturbine.C [hunk 2]
```diff
@@ -1274,41 +1285,43 @@ fsiTurbine::computeHubForceMomentForPart(
...
-        fsiForceNode[i] = pressureForceNode[i] + viscForceNode[i];
-        tmpMeshPos[i] = xyz[i] + xyz_disp[i];
-      }
-      computeEffForceMoment(
-        fsiForceNode.data(), tmpMeshPos.data(), l_hubForceMoment.data(),
-        hubPos.data());
-    }
-  }
+  // TODO: This is looping over the wrong buckets - Nodes instead of faces
+  // Is this even required anymore? Probly can delete
```
Comment (ashesh2512) [inline src/aero/fsi/FSIturbine.C:1300]: does this need to be deleted?

File: src/aero/fsi/FSIturbine.C [hunk 3]
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
...
-        // << 180.0*4.0*std::atan(0.25*totDispNode[3])/M_PI << ", " <<
-        // totDispNode[4] << ", " << totDispNode[5] << std::endl;
-
         // Now transfer the interpolated displacement to the CFD mesh node
-        computeBladeDisplacement(
-          totDispNode.data(), tmpNodePos.data(), dx, oldxyz,
-          brFSIdata_.bld_pitch[iBlade],
-          &brFSIdata_.bld_root_def[iBlade * 6 + 3],
-          brFSIdata_.bld_rloc[*dispMapNode + iStart]);
+        computeDisplacement(totDispNode.data(), tmpNodePos.data(), dx, oldxyz);
```
Comment (ashesh2512) [inline src/aero/fsi/FSIturbine.C:1937]: does this need to be deleted?

File: src/aero/fsi/FSIturbine.C [hunk 4]
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
...
+        //       brFSIdata_.bld_pitch[iBlade],
+        //       &brFSIdata_.bld_root_def[iBlade * 6 + 3],
+        //       brFSIdata_.bld_rloc[*dispMapNode + iStart]
```
Comment (ashesh2512) [inline src/aero/fsi/FSIturbine.C:1940]: does this need to be deleted?

File: src/aero/fsi/FSIturbine.C [hunk 5]
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
...
 
-    // if (!bulk_.parallel_rank()) {
-    //     std::cout << "Error in displacement for blade " << iBlade << " = " <<
-    //     errorNorm[0] << " " << errorNorm[1] << " " << errorNorm[2] <<
-    //     std::endl ;
-    // }
+    // NaluEnv::self().naluOutputP0() << "Error in displacement for blade " <<
+    // iBlade << " = " <<
+    //    errorNorm[0] << " " << errorNorm[1] << " " << errorNorm[2] <<
+    //    std::endl ;
```
Comment (ashesh2512) [inline src/aero/fsi/FSIturbine.C:1951]: does this need to be deleted?

File: src/aero/fsi/FSIturbine.C [hunk 6]
```diff
@@ -1968,46 +1933,37 @@ fsiTurbine::mapDisplacements()
...
 
-    // if (!bulk_.parallel_rank()) {
-    //     std::cout << "Error in velocity for blade " << iBlade << " = " <<
-    //     errorNorm[0] << " " << errorNorm[1] << " " << errorNorm[2] <<
-    //     std::endl ;
-    // }
+    // NaluEnv::self().naluOutputP0() << "Error in velocity for blade " <<
+    // iBlade << " = " <<
+    //    errorNorm[0] << " " << errorNorm[1] << " " << errorNorm[2] <<
+    //    std::endl ;
```
Comment (ashesh2512) [inline src/aero/fsi/FSIturbine.C:1959]: does this need to be deleted?

File: src/aero/fsi/FSIturbine.C [hunk 7]
```diff
@@ -2717,31 +2640,31 @@ fsiTurbine::computeLoadMapping()
...
   //         brFSIdata_.bld_rloc[iStart+i+1]); bld_dr_[iStart + i] =
   //         (bld_rmm_[iStart + i][1]-bld_rmm_[iStart + i][0]);
-  //         // if (!bulk_.parallel_rank())
-  //         //     std::cout << "i = " << i << ", " << bld_rmm_[iStart+i][0] <<
-  //         " - " << bld_rmm_[iStart+i][1] << ", dr = " << bld_dr_[iStart+i] <<
-  //         std::endl ;
+  //         NaluEnv::self().naluOutputP0() << "i = " << i << ", " <<
+  //         bld_rmm_[iStart+i][0] <<
+  //               " - " << bld_rmm_[iStart+i][1] << ", dr = " <<
+  //               bld_dr_[iStart+i] << std::endl ;
```
Comment (ashesh2512) [inline src/aero/fsi/FSIturbine.C:2656]: Is this left over code?

**Comment by ashesh2512** [comment]
Looks good to me other than a bunch of commented out code in `FSITurbine.C`. If that is intentional, I am good with this PR.

**Comment by psakievich** [comment]
@gantech is the commented code part of your testing infrastructure or can we delete it?

**Review by ashesh2512 — APPROVED** [review]

**Comment by gantech** [comment]
Thanks Ashesh.. I'm addressing your comments now

File: src/Realm.C [hunk 8]
```diff
@@ -3427,15 +3441,31 @@ Realm::populate_restart(double& timeStepNm1, int& timeStepCount)
...
+    if (does_mesh_move()) {
+
       // Redo all the mesh and motionAlg setup after reading files from the
       // restart reset the current_coordinate and mesh_velocity fields after
       // reading them
       init_current_coordinates();
+
       // reset the current time for the meshMotionAlgs
-      meshMotionAlg_->restart_reinit(foundRestartTime);
+      if (meshMotionAlg_)
```
[the remark this answers is not in the document]
Comment (gantech) [reply src/Realm.C:3453]: Agreed!

File: src/Realm.C [hunk 9]
```diff
@@ -3427,15 +3441,31 @@ Realm::populate_restart(double& timeStepNm1, int& timeStepCount)
...
+          auto part_vec = aeroModels_->fsi_parts();
+          for (auto* target_part : part_vec)
+            set_current_coordinates(target_part);
+        }
+      }
+
       compute_geometry();
-      meshMotionAlg_->post_compute_geometry();
+
+      if (meshMotionAlg_)
```
[the remark this answers is not in the document]
Comment (gantech) [reply src/Realm.C:3467]: Agreed!

File: src/aero/AeroContainer.C [hunk 10]
```diff
@@ -44,14 +44,14 @@ AeroContainer::AeroContainer(const YAML::Node& node)
         "look_ahead_and_create::error: Too many actuator line blocks");
     actuatorModel_.parse(*foundActuator[0]);
   }
-  std::vector<const YAML::Node*> foundFsi;
-  NaluParsingHelper::find_nodes_given_key("openfast_fsi", node, foundFsi);
-  if (foundFsi.size() > 0) {
+  // std::vector<const YAML::Node*> foundFsi;
+  // NaluParsingHelper::find_nodes_given_key("openfast_fsi", node, foundFsi);
```
[the remark this answers is not in the document]
Comment (gantech) [reply src/aero/AeroContainer.C:48]: Yeah.. This code should've worked. I'd like to leave it in there if we want to fix this later

File: src/aero/fsi/FSIturbine.C (hunk 3 above)
Comment (gantech) [reply src/aero/fsi/FSIturbine.C:1937]: Agreed. Will delete

**Comment by psakievich** [comment]
@ashesh2512 I'm going to merge after this CI passes. I will continue to address these issues in #1098 .  We're going to try and merge that soon after this PR.

**Pull request merged**